shared_model/cryptography/crypto_provider/crypto_verifier.cpp

```diff
@@ -0,0 +1,87 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "cryptography/crypto_provider/crypto_verifier.hpp"
+
+#include "common/hexutils.hpp"
+#include "common/result.hpp"
+#include "cryptography/ed25519_sha3_impl/crypto_provider.hpp"
+#include "interfaces/common_objects/byte_range.hpp"
+#include "multihash/multihash.hpp"
+#include "multihash/type.hpp"
+
+#if defined(USE_LIBURSA)
+#include "cryptography/ed25519_ursa_impl/crypto_provider.hpp"
+#define ED25519_PROVIDER CryptoProviderEd25519Ursa
+#endif
+
+using namespace iroha::expected;
+using namespace shared_model::crypto;
+using namespace shared_model::interface::types;
+
+using iroha::multihash::Multihash;
+
+using DefaultVerifier = CryptoProviderEd25519Sha3;
+
+namespace {
+  inline Result<bool, const char *> verifyMultihash(
+      ByteRange signature, const Blob &source, const Multihash &public_key) {
```

Comment:
this function does not store this object, it only *views* it, so it should be safe by design. then, this second level of type introspection is not very comfortable. to understand that it is ok to pass by value, you have to inspect 2 levels of types. and suppose if some of these levels changes to include more heavy data. it might get difficult to spot all such places where value is copied. i would leave it as is.

Reply:
OK, I think it can be moved to a separate discussion later.